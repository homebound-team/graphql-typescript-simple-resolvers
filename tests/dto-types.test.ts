import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Data Transfer Objects (DTOs)", () => {
  const schemaWithDTOs = `
    scalar Date

    enum Status {
      ACTIVE
      INACTIVE
    }

    # This is a mapped entity type
    type User {
      id: ID!
      name: String!
      email: String!
      createdAt: Date!
    }

    # These are DTO types (not mapped)
    type UserSummary {
      totalUsers: Int!
      activeUsers: Int!
      inactiveUsers: Int!
      lastUpdated: Date
    }

    type UserProfile {
      name: String!
      email: String!
      status: Status!
      profilePicture: String
      bio: String
      joinDate: Date
      followers: Int
      following: Int
    }

    type UserStats {
      userId: ID!
      postsCount: Int!
      commentsCount: Int!
      likesReceived: Int!
      averageRating: Float
      isVerified: Boolean!
      badges: [String!]!
      achievements: [String]
    }

    # Nested DTO
    type Address {
      street: String!
      city: String!
      state: String
      zipCode: String!
      country: String!
    }

    type UserDetails {
      profile: UserProfile!
      address: Address
      emergencyContact: UserProfile
    }

    type Query {
      # Returns mapped entity
      user(id: ID!): User

      # Returns DTOs
      userSummary: UserSummary!
      userProfile(id: ID!): UserProfile
      userStats(id: ID!): UserStats
      userDetails(id: ID!): UserDetails
    }
  `;

  it("generates DTO types with proper structure", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      mappers: {
        User: "./entities#UserEntity",
      },
      scalars: {
        Date: "Date",
      },
      enumValues: {},
    });

    // DTOs should be generated as full objects
    expect(code).toContain(`export interface UserSummary {
  totalUsers: number;
  activeUsers: number;
  inactiveUsers: number;
  lastUpdated: Date | null | undefined;
}`);
  });

  it("handles DTO fields with proper nullability", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      scalars: { Date: "Date" },
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface UserProfile {
  name: string;
  email: string;
  status: Status;
  profilePicture: string | null | undefined;
  bio: string | null | undefined;
  joinDate: Date | null | undefined;
  followers: number | null | undefined;
  following: number | null | undefined;
}`);
  });

  it("handles DTO fields with various scalar types", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface UserStats {
  userId: string;
  postsCount: number;
  commentsCount: number;
  likesReceived: number;
  averageRating: number | null | undefined;
  isVerified: boolean;
  badges: string[];
  achievements: Array<string | null | undefined> | null | undefined;
}`);
  });

  it("handles nested DTO types", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface UserDetails {
  profile: UserProfile;
  address: Address | null | undefined;
  emergencyContact: UserProfile | null | undefined;
}`);

    expect(code).toContain(`export interface Address {
  street: string;
  city: string;
  state: string | null | undefined;
  zipCode: string;
  country: string;
}`);
  });

  it("generates resolvers for all types including DTOs", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      mappers: {
        User: "./entities#UserEntity",
      },
      scalars: {},
      enumValues: {},
    });

    // Should generate resolvers for mapped types
    expect(code).toContain("UserResolvers");

    // Should also generate resolver types for DTOs
    expect(code).toContain("UserSummaryResolvers");
    expect(code).toContain("UserProfileResolvers");
    expect(code).toContain("UserStatsResolvers");
    expect(code).toContain("AddressResolvers");
  });

  it("generates query resolvers that return DTOs", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      mappers: {
        User: "./entities#UserEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain("userSummary: Resolver<{}, {}, UserSummary>;");
    expect(code).toContain("userProfile: Resolver<{}, QueryUserProfileArgs, UserProfile | null | undefined>;");
  });

  it("snapshots DTO generation", async () => {
    const schema = createTestSchema(schemaWithDTOs);
    const code = await runPlugin(schema, {
      mappers: {
        User: "./entities#UserEntity",
      },
      scalars: {
        Date: "Date",
      },
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });
});
