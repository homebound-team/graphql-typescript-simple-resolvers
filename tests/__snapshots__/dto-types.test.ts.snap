// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Data Transfer Objects (DTOs) snapshots DTO generation 1`] = `
"import { GraphQLResolveInfo, GraphQLScalarType } from "graphql";
import { Context } from "./context";
import { UserEntity } from "./entities";

export interface Resolvers {
  User: UserResolvers;
  Query: QueryResolvers;
  UserSummary?: UserSummaryResolvers;
  UserProfile?: UserProfileResolvers;
  UserStats?: UserStatsResolvers;
  Address?: AddressResolvers;
  UserDetails?: UserDetailsResolvers;
  Date: GraphQLScalarType;
}

export type UnionResolvers = {};

export interface UserResolvers {
  id: Resolver<UserEntity, {}, string>;
  name: Resolver<UserEntity, {}, string>;
  email: Resolver<UserEntity, {}, string>;
  createdAt: Resolver<UserEntity, {}, Date>;
}

export interface QueryResolvers {
  user: Resolver<{}, QueryUserArgs, UserEntity | null | undefined>;
  userSummary: Resolver<{}, {}, UserSummary>;
  userProfile: Resolver<{}, QueryUserProfileArgs, UserProfile | null | undefined>;
  userStats: Resolver<{}, QueryUserStatsArgs, UserStats | null | undefined>;
  userDetails: Resolver<{}, QueryUserDetailsArgs, UserDetails | null | undefined>;
}

export interface UserSummaryResolvers {
  totalUsers: Resolver<UserSummary, {}, number>;
  activeUsers: Resolver<UserSummary, {}, number>;
  inactiveUsers: Resolver<UserSummary, {}, number>;
  lastUpdated: Resolver<UserSummary, {}, Date | null | undefined>;
}

export interface UserProfileResolvers {
  name: Resolver<UserProfile, {}, string>;
  email: Resolver<UserProfile, {}, string>;
  status: Resolver<UserProfile, {}, Status>;
  profilePicture: Resolver<UserProfile, {}, string | null | undefined>;
  bio: Resolver<UserProfile, {}, string | null | undefined>;
  joinDate: Resolver<UserProfile, {}, Date | null | undefined>;
  followers: Resolver<UserProfile, {}, number | null | undefined>;
  following: Resolver<UserProfile, {}, number | null | undefined>;
}

export interface UserStatsResolvers {
  userId: Resolver<UserStats, {}, string>;
  postsCount: Resolver<UserStats, {}, number>;
  commentsCount: Resolver<UserStats, {}, number>;
  likesReceived: Resolver<UserStats, {}, number>;
  averageRating: Resolver<UserStats, {}, number | null | undefined>;
  isVerified: Resolver<UserStats, {}, boolean>;
  badges: Resolver<UserStats, {}, readonly string[]>;
  achievements: Resolver<UserStats, {}, ReadonlyArray<string | null | undefined> | null | undefined>;
}

export interface AddressResolvers {
  street: Resolver<Address, {}, string>;
  city: Resolver<Address, {}, string>;
  state: Resolver<Address, {}, string | null | undefined>;
  zipCode: Resolver<Address, {}, string>;
  country: Resolver<Address, {}, string>;
}

export interface UserDetailsResolvers {
  profile: Resolver<UserDetails, {}, UserProfile>;
  address: Resolver<UserDetails, {}, Address | null | undefined>;
  emergencyContact: Resolver<UserDetails, {}, UserProfile | null | undefined>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QueryUserArgs {
  id: string;
}
export interface QueryUserProfileArgs {
  id: string;
}
export interface QueryUserStatsArgs {
  id: string;
}
export interface QueryUserDetailsArgs {
  id: string;
}
export interface UserSummary {
  totalUsers: number;
  activeUsers: number;
  inactiveUsers: number;
  lastUpdated: Date | null | undefined;
}

export interface UserProfile {
  name: string;
  email: string;
  status: Status;
  profilePicture: string | null | undefined;
  bio: string | null | undefined;
  joinDate: Date | null | undefined;
  followers: number | null | undefined;
  following: number | null | undefined;
}

export interface UserStats {
  userId: string;
  postsCount: number;
  commentsCount: number;
  likesReceived: number;
  averageRating: number | null | undefined;
  isVerified: boolean;
  badges: string[];
  achievements: Array<string | null | undefined> | null | undefined;
}

export interface Address {
  street: string;
  city: string;
  state: string | null | undefined;
  zipCode: string;
  country: string;
}

export interface UserDetails {
  profile: UserProfile;
  address: Address | null | undefined;
  emergencyContact: UserProfile | null | undefined;
}

export enum Status {
  Active = "ACTIVE",
  Inactive = "INACTIVE",
}
"
`;
